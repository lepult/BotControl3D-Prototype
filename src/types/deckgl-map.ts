// region Imports
import { Position, Color, TransitionInterpolator } from '@deck.gl/core/typed';
import { IconLayer, PathLayer, PolygonLayer } from '@deck.gl/layers/typed';
import { ScenegraphLayer } from '@deck.gl/mesh-layers/typed';
import { IMergedMapElement, IZone } from './pudu-api/robotMap';
import { CustomDestinationType } from './api/destination';
// endregion

export interface IIconData extends IMergedMapElement {
    name: string;
    position: Position;
    color?: Color;
    angle?: number;
    selected?: boolean;
    routeData: {
        isRouteDestination: boolean;
        isNextDestination: boolean;
        isEarlierDestination: boolean;
        isFinalDestination: boolean;
        isPreviousDestination: boolean;
    },
    customType?: CustomDestinationType;
}

export interface IPathData extends IMergedMapElement {
    name: string;
    path: Array<number>;
    color?: Color;
}

export type ISceneGraphData = {
    name: string;
    position: [number, number];
    angle: number;
    color?: Color;
};

export interface IPolygonData extends IZone {
    name: string;
    color?: Color;
}

export type TTooltip = {
    picked?: boolean;
    object?: IIconData & IPathData & ISceneGraphData & IPolygonData;
    x?: number;
    y?: number;
};

export type TLayers = (IconLayer<IIconData, NonNullable<unknown>>
    | PathLayer<IPathData, NonNullable<unknown>>
    | ScenegraphLayer<ISceneGraphData, NonNullable<unknown>>
    | PolygonLayer<IPolygonData, NonNullable<unknown>>)[];

export type TViewState = {
    bearing: number,
    latitude: number,
    longitude: number,
    pitch: number,
    zoom: number,
    maxZoom?: number,
    minZoom?: number,
    rotationX?: number,
    maxPitch?: number,
    minPitch?: number,
    transitionDuration?: number,
    transitionInterpolator?: TransitionInterpolator,
}

export enum MapRobotStatus {
    'Offline' = 'Offline',
    'Idle' = 'Idle',
    'Charged' = 'Charged',
    'Charging' = 'Charging',
    'ChargingError' = 'ChargingError',
    'ArrivedAtDiningOutlet' = 'ArrivedAtDiningOutlet',
    'SendToChargingPile' = 'SendToChargingPile',
    'WaitForDoor' = 'WaitForDoor',
    'WaitForElevator' = 'WaitForElevator',
    'MovingWithElevator' = 'MovingWithElevator',
    'ArrivedAtPickupDestination' = 'ArrivedAtPickupDestination',
    'PrepareDriveToDestination' = 'PrepareDriveToDestination',
    'SendToDestination' = 'SendToDestination',
    'ArrivedAtDestination' = 'ArrivedAtDestination',
    'Avoid' = 'Avoid',
    'Blocked' = 'Blocked',
    'Cancel' = 'Cancel',
    'Error' = 'Error',
    'Pause' = 'Pause',
}

export enum PreviewType {
    Robot,
    Floor,
}

export enum DragMode {
    translate = 'translate',
    rotate = 'rotate',
}

export type TUndoStackItem = {
    id: string,
    position: [number, number, number],
    orientation: [number, number, number],
}

export type TRobotLayerData = {
    name: string,
    robotId: string,
    position: [number, number],
    icon: string,
    color: Color,
    orientation: [number, number, number],
};
