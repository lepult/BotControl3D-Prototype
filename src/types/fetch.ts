export enum FetchState {
    initial,
    pending,
    fulfilled,
    rejected,
}
