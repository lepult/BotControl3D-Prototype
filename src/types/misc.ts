export enum UserType {
    guest,
    team,
    admin,
}
