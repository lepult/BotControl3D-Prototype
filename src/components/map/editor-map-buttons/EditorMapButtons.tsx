import React, { FC } from 'react';
import { ModelType } from '../../../constants/hardcoded-data/models';
import { TViewState } from '../../../types/deckgl-map';
import '../buttons.scss';
import ImportButton from './buttons/ImportButton';
import ChangeInitialViewButton from './buttons/ChangeInitialViewButton';
import SaveButton from './buttons/SaveButton';
import ResetViewButton from './buttons/ResetViewButton';
import CancelButton from './buttons/CancelButton';

const EditorMapButtons: FC<{
    floorModels: ModelType[],
    viewState: TViewState | undefined,
    mapId: number,
}> = ({
    floorModels,
    viewState,
    mapId,
}) => (
    <div className="user-mode-buttons__wrapper">
        <div
            style={{
                position: 'relative',
                width: '100%',
                height: '100%',
            }}
        >
            <div
                className="map-buttons"
                style={{
                    alignItems: 'end',
                    top: 0,
                    left: 0,
                    right: 0,
                    padding: '10px',
                }}
            >
                <ImportButton/>
                <ResetViewButton/>
                <ChangeInitialViewButton viewState={viewState} mapId={mapId}/>
            </div>
            <div
                className="map-buttons"
                style={{
                    bottom: '10px',
                    left: '50%',
                    transform: 'translateX(-50%)',
                    display: 'flex',
                    flexDirection: 'row',
                }}
            >
                <SaveButton floorModels={floorModels}/>
                <div style={{ marginLeft: '10px' }}>
                    <CancelButton/>
                </div>
            </div>
        </div>
    </div>

);

export default EditorMapButtons;
