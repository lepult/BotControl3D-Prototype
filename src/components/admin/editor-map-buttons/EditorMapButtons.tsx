import React, { FC } from 'react';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { Button } from 'chayns-components';
import { useDispatch } from 'react-redux';
import { changeAdminModeType, resetViewState } from '../../../redux-modules/misc/actions';
import { changeInitialViewState } from '../../../redux-modules/map/actions';
import { AdminModeType } from '../../../types/misc';
import { ModelType } from '../../../constants/models';
import { TViewState } from '../../../types/deckgl-map';
import { createDialog, DialogButtonType, DialogType } from 'chayns-api';

const EditorMapButtons: FC<{
    floorModels: ModelType[],
    viewState: TViewState | undefined,
    mapId: number,
}> = ({
    floorModels,
    viewState,
    mapId,
}) => {
    const dispatch = useDispatch();

    return (
        <div>
            <Button
                onClick={() => {
                    const input = document.createElement('input');
                    input.type = 'file';
                    input.accept = '.glb';
                    input.onchange = _ => {
                        console.log('FileList', input.files);
                        console.log('Array', Array.prototype.slice.call(input.files));
                        // TODO Upload to chayns Space.
                        void createDialog({
                            type: DialogType.ALERT,
                            text: "Das Importieren von Dateien ist nicht implementiert.",
                            buttons: [{ type: DialogButtonType.CANCEL, text: 'Schließen' }],
                        }).open();
                    };
                    input.click();
                }}
            >
                Import
            </Button>
            <Button
                onClick={() => dispatch(resetViewState())}
            >
                Reset View
            </Button>
            <Button
                onClick={() => {
                    const newInitialViewState = {
                        bearing: viewState?.bearing || 0,
                        latitude: viewState?.latitude || 0,
                        longitude: viewState?.longitude || 0,
                        pitch: viewState?.pitch || 0,
                        zoom: viewState?.zoom || 21,
                    };
                    console.log('newInitialViewState', newInitialViewState);
                    dispatch(changeInitialViewState({
                        mapId,
                        viewState: newInitialViewState,
                    }))
                }}
            >
                Initialen ViewState setzen
            </Button>
            <Button
                onClick={() => {
                    console.log('Models', floorModels);
                    dispatch(changeAdminModeType({ adminModeType: AdminModeType.default }));
                    void createDialog({
                        type: DialogType.ALERT,
                        text: "Das Speichern der Modelle ist nicht implementiert. Mit dem Neuladen der Seite werden die Positionen zurückgesetzt.",
                        buttons: [{ type: DialogButtonType.CANCEL, text: 'Schließen' }],
                    }).open();
                }}
            >
                Speichern
            </Button>
        </div>
    )
};

export default EditorMapButtons;
