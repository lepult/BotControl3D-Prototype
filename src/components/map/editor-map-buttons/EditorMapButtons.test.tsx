import React from 'react';
import renderer from 'react-test-renderer';
import EditorMapButtons from './EditorMapButtons';
import { TViewState } from '../../../types/deckgl-map';

jest.mock('./buttons/ImportButton', () => () => <button className="import"/>);
jest.mock('./buttons/ChangeInitialViewButton', () => (
    { viewState, mapId }: { viewState: TViewState, mapId: number }
) => <button className={viewState && mapId ? 'change-initial-view' : ''}/>);
jest.mock('./buttons/SaveButton', () => (
    { floorModels }: { floorModels: number[] }
) => <button className={floorModels?.length > 0 ? 'save' : ''} />);
jest.mock('./buttons/CancelButton', () => () => <button className="cancel"/>);
jest.mock('./buttons/ShortcutsButton', () => () => <button className="shortcuts"/>);
jest.mock('../user-mode-buttons/interaction-buttons/ResetViewButton', () => () => <button className="reset"/>);

it('EditorMapButtons should result in showing all buttons', () => {
    const component = renderer.create(
        <EditorMapButtons
            floorModels={[{ id: '1', url: '', position: [0, 0, 0], orientation: [0, 0, 0]}]}
            viewState={{ bearing: 0, latitude: 0, longitude: 0, pitch: 0, zoom: 0 }}
            mapId={1}
        />,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})