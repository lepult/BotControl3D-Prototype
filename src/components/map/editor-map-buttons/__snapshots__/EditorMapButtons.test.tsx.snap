// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorMapButtons should result in showing all buttons 1`] = `
<div
  className="user-mode-buttons__wrapper"
>
  <div
    style={
      {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "left": 0,
          "padding": "10px",
          "right": 0,
          "top": 0,
        }
      }
    >
      <button
        className="change-initial-view"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "alignItems": "end",
          "bottom": 0,
          "padding": "10px",
          "right": 0,
        }
      }
    >
      <button
        className="import"
      />
      <button
        className="shortcuts"
      />
      <button
        className="reset"
      />
    </div>
    <div
      className="map-buttons"
      style={
        {
          "bottom": "10px",
          "display": "flex",
          "flexDirection": "row",
          "left": "50%",
          "transform": "translateX(-50%)",
        }
      }
    >
      <button
        className="save"
      />
      <div
        style={
          {
            "marginLeft": "10px",
          }
        }
      >
        <button
          className="cancel"
        />
      </div>
    </div>
  </div>
</div>
`;
